docs/RFCS/20211018_index_recommendation.md

```diff
@@ -0,0 +1,207 @@
+In a similar way, if the addition of our hypothetical indexes caused some existing indexes to become unused or rarely used, we would recommend that these indexes be deleted.
+If the index is still occasionally used, we need to ensure that removing it doesn't negatively affect overall performance. A good way to check this could be contrasting the writes to this index with the reads.
+As in, if an index is read from more than it is written to, it's better than the opposite. To fully ensure that this has not caused a regression however, we should re-run *W* with the index hypothetically dropped.
+Additionally, before we delete an index, we should ensure that no queries are regularly using hints to force that index.
+After this step, we have our optimal *X* that will be recommended to the user.
+
+In terms of the final output, we will definitely have recommended indexes. In addition, we should have a quantifiable "impact score" associate with an index recommendation that we can use to justify why the index would be beneficial to users.
+We can also include further information with this, such as what queries are affected and/or affected the most. For "drop" recommendations, we should have a similar metric.
+If an index is being frequently written to and infrequently accessed, the score would be higher than if only the latter was true.
+
+### Issues and Considerations
+One issue with this approach is that it can become a feedback loop where adding new indexes affects the existing ones, so we remove them, and then that allows for new potential indexes to be useful.
+This is a tradeoff that must be accepted - otherwise the algorithm could run infinitely. Implementation will begin with recommending potential indexes to add, followed by recommending indexes to remove.
+Another heuristic that could be added is random swapping of a small subset of indexes being tried with other indexes that were found in *W's* candidate sets. If the total cost of *W* is lower with this random swapping, we keep this configuration and continue as described above.
+The number of times this would be tried would be limited, to avoid having an inefficient algorithm. As an aside, we can also independently use index usage metrics to determine if there are any unused indexes that we should recommend be deleted.
+
+An additional issue is the lack of statistics on non-indexed tables. This will impact the plan that is chosen by the optimizer. Since statistics collection is a long and involved task, there is no clear way of mitigating this.
+Instead, this is a limitation that we must accept for now, especially since this will not stop us from making recommendations (it will just potentially impact their quality).
+
+We need to also factor in a cost for database writes in the costing algorithm, which is not done by the optimizer. For database reads, indexes can only have positive impact, whereas for writes, they can have a negative impact.
```

Comment:
ditto: I think it's sufficient to simply point out that there is some overhead for maintaining indexes, and that the current proposal does not solve for this. This write overhead is mentioned in the section above and in the section below - you might want to consolidate these into a single paragraph in "issues and considerations" or "drawbacks".